File: src/main/kotlin/com/huawei/cangjie/psi/CjUnaryExpression.java
package com.huawei.cangjie.psi;

import com.huawei.cangjie.CjNodeTypes;
import com.intellij.lang.ASTNode;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;

public abstract class CjUnaryExpression extends CjExpressionImpl implements CjOperationExpression {
    public CjUnaryExpression(ASTNode node) {
        super(node);
    }

    @Nullable
    @IfNotParsed
    public abstract CjExpression getBaseExpression();

    @Override
    @NotNull
    public CjSimpleNameExpression getOperationReference() {
        return  Objects.requireNonNull(findChildByType(CjNodeTypes.OPERATION_REFERENCE));
    }

    public IElementType getOperationToken() {
        return getOperationReference().getReferencedNameElementType();
    }
}

File: src/main/kotlin/com/linqingying/cangjie/psi/CjUnaryExpression.kt
package com.linqingying.cangjie.psi

import com.linqingying.cangjie.CjNodeTypes
import com.intellij.lang.ASTNode
import com.intellij.psi.tree.IElementType
import com.intellij.psi.util.PsiTreeUtil
import java.util.*

abstract class CjUnaryExpression(node: ASTNode) : CjExpressionImpl(node), CjOperationExpression {

    @get:IfNotParsed
    open val baseExpression: CjExpression? get() = PsiTreeUtil.getPrevSiblingOfType(operationReference, CjExpression::class.java)



    override val operationReference: CjSimpleNameExpression
        get() = findChildByType(CjNodeTypes.OPERATION_REFERENCE)!!
    val operationToken: IElementType
        get() = operationReference.getReferencedNameElementType()
}
